Add new tutorial for SAM3 (#6813)
galaxyproject/training-material: topics/ecology/tutorials/SAM3/tutorial.md

---
layout: tutorial_hands_on

> title: SAM3 – AI-based Semantic Segmentation of Images and Videos

Reviewer: ```suggestion
title: SAM3 – AI-based Semantic Segmentation of marine biodiversity Images and Videos
```

Author: I'm not sure about this change, bc for me this is not a tutorial specifically about marine biodiversity, I just use it as an example.

Reviewer: It seems to me 1/ it is MOOREV oriented and so marine biodiversity and 2/ it is good to let others communities creating their SAM3 tuto on other subject and so maybe giving a precision here can be good. BUT ;) Really not sure it is mandatory to change it, please don't hesitate to not modify it!

Author: Argument accepted

questions:

> - How can you automatically segment objects in images or videos using a simple text prompt?

Reviewer: ```suggestion
- How can you automatically segment marine biodiversity objects in images or videos using a simple text prompt?
```

objectives:
- Segment marine animals in a photograph and/or video using a text prompt
time_estimation: 20M
key_points:
- SAM3
- IA
- Segmentation

> - Semantic

Reviewer: ```suggestion
- Adding annotation to images and videos automatically through a prompt is something biodiversity data are waiting for
- Segment, identify and track animals is now something several Galaxy tools can help to do
- SAM3 Galaxy tool allows user to do it using a prompt
```

tags:
  - Marine ecosystems

>   - biodiversity

Reviewer: ```suggestion
  - Biodiversity
```

  - AI segmentation
contributions:
  authorship:

>     - TuturBaba

Reviewer: ```suggestion
    - TuturBaba
    - yvanlebras
    - nadinelebris
```

  funding:
    - Moorev
    - sorbonneuniv
    - ISYEB

>     - mnhn

Reviewer: ```suggestion
    - mnhn
    - pndb    
```

lang: en
translations:
    - fr
---

This tutorial will guide you through using the {% tool [SAM3](toolshed.g2.bx.psu.edu/repos/ecology/sam3_semantic_segmentation/sam3_semantic_segmentation/1.0.1+galaxy4) %} (Segment Anything Model 3) tool on Galaxy. SAM3 can automatically detect and segment objects in images or videos using text prompts, with no specific training required.

We will work through two concrete examples from the Moorev project:
1. A **photograph of a jellyfish** (*Pelagia noctiluca*)
2. A **video of shrimps**

> <agenda-title>In this tutorial, we will cover:</agenda-title>
>
> 1. TOC
> {:toc}
>
{: .agenda}

> <details-title>Quick introduction to how Galaxy works</details-title>
>

> > You can leave the tutorial and return to the main screen by clicking outside the tutorial window.

Reviewer: It is only for people using the GTN webhook from within usegalaxy.eu (and some other Galaxy instance) but not all... So maybe remove this part no ?

Reviewer: ..or maybe make thiese 2 sentences something "optional", only for puser using Galaxy webhook?

> > You can resume where you left off at any time by clicking {% icon level %}.

Reviewer: same here. Ideally, it would be good to have these elements displayed when the tuto is displayed from Galaxy instance webhook. Not sure it is possible / easy to add...

>
> **Log in to Galaxy**

> > 1. Open your preferred web browser (Chrome, Safari, or Firefox — not Edge!)

Reviewer: I am personnaly quite ok with this but in fact it seems that many users are using edge, and it seems to work not so badly isn't it?

Author: > <details-title>Quick introduction to how Galaxy works</details-title>
I reused the 'champ block' tutorial for this section, so I didn't write it myself :)

Author: For this bloc: `<details-title>Quick introduction to how Galaxy works</details-title>`
I reused the 'champ block' tutorial for this section, so I didn't write it myself :)

Reviewer: ok, I think beter to avoid special mention of edge, and maybe we / I have to make a new pass on "champ block" tuto to clean / update it ;)

> > 2. Go to your Galaxy instance

Reviewer: ```suggestion
> 2. Go to your Galaxy instance (please verify that the Galaxy instance you want to use propose SAM3 tool as Galaxy Europe is doing)
```

> 3. Log in or create an account
>
> ![Screenshot of the Galaxy interface with the login button highlighted](../../images/SAM3/galaxy_homepage.png)
>
> This screenshot shows the Galaxy Ecology instance, available at [usegalaxy.eu](https://ecology.usegalaxy.eu/)
>
> The Galaxy home page is divided into 3 panels:
> * **Tools** on the left
> * The **visualisation panel** in the centre
> * The **history** of analyses and files on the right
>
> ![Screenshot of the Galaxy interface showing the three panels](../../images/champbloc/galaxy_interface.png)
>
> The first time you use Galaxy, there will be no files in your history panel.
{: .details}

# Loading data into Galaxy

> Before running SAM3, you need to import the following files into Galaxy:

Reviewer: ```suggestion
Before running SAM3 Galaxy tool, you need to import the following files into Galaxy:
```

- The jellyfish photo: `https://zenodo.org/records/19890809/files/Moorev-jellyfish.jpg?download=1`
- The shrimp video: `https://zenodo.org/records/19891364/files/2024-09-20-PorzBreign-shrimps.mp4?download=1`

> <tip-title>Upload data to Galaxy</tip-title>
> * Copy the link location
> * 1- Click {% icon galaxy-upload %} **Upload** at the top of the activity panel
> * 2- Select {% icon galaxy-wf-edit %} **Paste/Fetch Data**
> * 3- Paste the link(s) into the text field `https://zenodo.org/records/19890809/files/Moorev-jellyfish.jpg`
> * 4- Press **Start**
> * 5- **Close** the window
> ![Galaxy upload link](../../images/SAM3/upload_by_link.gif){: style="width:80%; display:block; margin:auto;"}
{: .tip}

# Segmenting an image: the jellyfish photograph

> In this first section, we will run SAM3 on the photo `Moorev-jellyfish.jpg` to detect and segment the jellyfish.

Reviewer: ```suggestion
In this first section, we will run SAM3 Galaxy tool on the photo `Moorev-jellyfish.jpg` to detect and segment the jellyfish.
```

>    >
>    > Avoid overly vague descriptions like `animal` if you are specifically looking for a jellyfish.
>    > You can also use more descriptive prompt like `small blue fish`, but results may vary depending on the objects you want to detect.
>    {: .tip}
>
> 2. Click **Run Tool**
>
>    > <comment-title>Processing time</comment-title>
>    >
>    > Processing may take a few minutes depending on the image size and the resources available on the server. Wait until the outputs appear in green in the history.
>    >
>    {: .comment}
>
> 3. Once processing is complete, the following outputs appear in your history:
>    - **COCO Annotation**: the `annotations.json` file containing the segmentation masks
>    - **Annotated Outputs**: the collection of annotated images with overlaid masks
>
> 4. Viewing the annotated result
>
>    You should see the jellyfish outlined with a coloured mask and a bounding box.
>
>    > <tip-title>Display the image in the history</tip-title>
>    >
>    > Click on **Annotated Outputs** in the history panel:
>    > ![Click on Annotated Outputs](../../images/SAM3/history-annoted-outputs.png){: style="width:50%; display:block; margin:auto;"}
>    >

> >    > Then click the {% icon galaxy-eye %} icon to display the image in the central panel:

Reviewer: ```suggestion
>    > Then use the {% icon galaxy-eye %} icon to display the image in the central panel:
```

Reviewer: to limit the use of "click" terms ;)

>    > ![Click the eye icon to display](../../images/SAM3/history-clic-eye.png){: style="width:50%; display:block; margin:auto;"}
>    >
>    > Or click {% icon galaxy-save %} to download the file directly.
>    {: .tip}
>
>    ![SAM3 segmentation mask on the jellyfish photo](../../images/SAM3/display_picture.png "Segmentation result"){: style="width:75%; display:block; margin:auto;"}
>
> 5. Exploring the COCO file

> >    - Look at the content of your **COCO Annotation** in your history

Reviewer: ```suggestion
>    - Look at the content of your **COCO Annotation** file in your history
```

> >    - Click {% icon galaxy-eye %} to view the JSON, or {% icon galaxy-save %} to download it

Reviewer: ```suggestion
>    - Use {% icon galaxy-eye %} to view the JSON, or {% icon galaxy-save %} to download it
```

>
>    The file contains the `images`, `annotations`, and `categories` fields. Each annotation includes:
>    - `segmentation`: the polygon coordinates of the mask
>    - `bbox`: the bounding box `[x, y, width, height]`
>    - `category_id`: the identifier of the detected class (`1` = `jellyfish`)
>
>    > <details-title>Export to YOLO format (optional)</details-title>
>    >
>    > If you need to train a YOLO model with your annotations, you can export results in YOLO format in addition to COCO.
>    > In the {% icon param-select %} *"Output formats"* parameter, select `COCO` **and/or** `YOLO segmentation masks` **and/or** `YOLO bounding boxes`.
>    >
>    > > <details-title>YOLO segmentation format</details-title>
>    > >
>    > > Each line in a YOLO segmentation label file follows this format:
>    > > ```
>    > > <class_id> <x1> <y1> <x2> <y2> ... <xn> <yn>
>    > > ```
>    > > Coordinates are normalised between 0 and 1 relative to the image dimensions.
>    > > Example for a jellyfish (class 0): `0 0.423 0.312 0.456 0.298 ...`
>    > >
>    > {: .details}
>    >
>    {: .details}
>
{: .hands_on}

# Segmenting a video: the shrimp video

> In this second section, we will apply SAM3 to the video `2024-09-20-PorzBreign-shrimps.mp4`. SAM3 analyses the video frame by frame, tracking the shrimps over time.

Reviewer: ```suggestion
In this second section, we will execute SAM3 tool to the video `2024-09-20-PorzBreign-shrimps.mp4`. SAM3 model analyses the video frame by frame, tracking the shrimps over time.
```

>    > Video processing takes significantly longer than processing a single image. For a video of a few minutes, expect between 5 and 20 minutes depending on the server and the stride chosen.
>    >
>    {: .comment}
>
> 3. The following outputs appear in your history:
>    - **COCO Annotation**: the JSON file with annotations for each processed frame
>    - **Annotated Outputs**: the annotated video with segmentation masks overlaid frame by frame
>
> 4. Viewing the annotated video
>
>    - Click on **Annotated Outputs** in the history panel
>    - Click {% icon galaxy-eye %}
>    - Click {% icon galaxy-visualise %}
>    - Select **Media Player**
>
>    > <tip-title>Display the video in Galaxy</tip-title>
>    >
>    > ![Display the annotated video in Galaxy](../../images/SAM3/display_video.png){: style="width:75%; display:block; margin:auto;"}
>    >
>    > > <warning-title>Video not loading?</warning-title>
>    > >
>    > > The video may not load in Galaxy for several reasons:
>    > > - The file is too large for your internet connection
>    > > - The {% icon param-select %} *"Video quality"*: `Original quality (copy)` setting makes in-browser playback unavailable
>    > >

> >    > > In that case, click {% icon galaxy-save %} to download the video and play it locally with your usual media player.

Reviewer: ```suggestion
>    > > In that case, use {% icon galaxy-save %} to download the video and play it locally with your usual media player.
```

> >    > > In that case, click {% icon galaxy-save %} to download the video and play it locally with your usual media player.

Reviewer: ```suggestion
>    > > In that case, click {% icon galaxy-save %} to download the video and play it locally with your local media player.
```

>    > {: .warning}
>    >
>    {: .tip}
>
>    You will see the shrimps tracked with a coloured segmentation mask throughout the video.
>
>    ![SAM3 segmentation of shrimps](../../images/SAM3/2024-09-20-PorzBreign-shrimps.gif){: style="width:50%; display:block; margin:auto;"}
>
> 5. Downloading the annotated video
>    - Click the **Annotated Outputs** collection

> >    - Click {% icon galaxy-save %} to download the `.mp4` video

Reviewer: ```suggestion
>    - Use {% icon galaxy-save %} to download the `.mp4` video
```

>
>    > <comment-title>Limitations of SAM3 and pre-processing</comment-title>
>    >

> >    > As you can see, SAM3 is far from perfect. Adjusting the confidence threshold can help, but it does not solve everything. Pre-processing your images or videos is often necessary to improve results.

Reviewer: ```suggestion
>    > SAM3 tool is a first attempt to propose prompt-based  Galaxy tool. As it is using SAM3 model, you can have highly heterogenous results in term of quality depending on the objects you are searching to segment, notably if such kind of object can be on data used to train the SAM3 model. Adjusting the confidence threshold can help, but it does not solve everything. Pre-processing your images or videos is often necessary to improve results.
```

Reviewer: Something like this no ?

> >    > To learn more, check out the dedicated tutorial: **Tuto Moorev**

Reviewer: there is such "Tuto Moorev" tuto? Or it is / it will use using a tag on GTN ?

Author: I was just mentioning that I want to create a tutorial for the Moorev project. For now I only picked this as a reference, but once my new tutorial is ready I will tag it on GTN.

>    >
>    {: .comment}
>
{: .hands_on}

# Conclusion

> You now know how to use SAM3 on Galaxy to:

Reviewer: ```suggestion
You now know how to use SAM3 Galaxy tool to:
```

>
> You can leave the tutorial and return to the main screen by clicking outside the tutorial window.
> You can resume where you left off at any time by clicking {% icon level %}.
>
> **Log in to Galaxy**
> 1. Open your preferred web browser (Chrome, Safari, or Firefox — not Edge!)
> 2. Go to your Galaxy instance
> 3. Log in or create an account
>
> ![Screenshot of the Galaxy interface with the login button highlighted](../../images/SAM3/galaxy_homepage.png)
>
> This screenshot shows the Galaxy Ecology instance, available at [usegalaxy.eu](https://ecology.usegalaxy.eu/)
>
> The Galaxy home page is divided into 3 panels:
> * **Tools** on the left
> * The **visualisation panel** in the centre
> * The **history** of analyses and files on the right
>
> ![Screenshot of the Galaxy interface showing the three panels](../../images/champbloc/galaxy_interface.png)
>
> The first time you use Galaxy, there will be no files in your history panel.
{: .details}

# Loading data into Galaxy

Before running SAM3, you need to import the following files into Galaxy:
- The jellyfish photo: [Moorev-jellyfish.jpg](https://zenodo.org/records/19890809/files/Moorev-jellyfish.jpg?download=1)

> - The shrimp video: [2024-09-20-PorzBreign-shrimps.mp4](https://zenodo.org/records/19891364/files/2024-09-20-PorzBreign-shrimps.mp4?download=1)

Reviewer 2: Could these links be put directly transparently to copy paste them (in paste and fetch for the upload of data). It would avoid having them locally.

Reviewer 2: ```suggestion
- The jellyfish photo: `https://zenodo.org/records/19890809/files/Moorev-jellyfish.jpg?download=1`
- The shrimp video: `https://zenodo.org/records/19891364/files/2024-09-20-PorzBreign-shrimps.mp4?download=1`
```

Reviewer 2: Something like that

> <tip-title>Upload data to Galaxy</tip-title>
>
> 1. Click the upload tool {% icon galaxy-upload %} at the top left of the tools panel
>
> 2. Select **Choose local files**
>    ![Galaxy upload window](../../images/champbloc/upload.png "Upload your data")
>
> 3. Browse your computer and select the two files:
>    - `Moorev-jellyfish.jpg`
>    - `2024-09-20-PorzBreign-shrimps.mp4`

> >

Reviewer 2: Thus this would need to be updated
But I think there are snippets existing for this like this one https://galaxyproject.github.io/training-material/faqs/galaxy/datasets_import_via_link.html

Author: Is it a deliberate choice to not include an image in the snippet? If so, can I propose a change?

Reviewer 2: I don't know if it's a deliberate choice, but I think an image can very be welcomed ! I think you can go ahead on proposing an update of the snippet, right @shiltemann ?

Reviewer 3: @TuturBaba yes, we are always happy if people are willing to improve the snippets, please feel free to include any updates to snippets you want/need as part of this PR. Thanks!

Author: @shiltemann Thank you! I also have a question: what is the file size limit?

Reviewer 3: Do you mean for input datasets? There is no hard limit, but try to make them as small as you can while not losing so much information that you do not get enough good results to illustrate the analysis and scientific content etc. It depends a bit on the tools and topic how small you can go.

As small as possible, as large as necessary I'd say

>
>    - {% icon param-file %} *"Model data"*: `Segment Anything Model 3 (SAM 3)` (default)
>    - {% icon param-select %} *"Input type"*: `One or more images` (default)
>    - {% icon param-file %} *"Input images"*: `Moorev-jellyfish.jpg`
>    - {% icon param-select %} *"Output formats"*: `COCO`
>    - {% icon param-text %} *"Text prompt"*: `jellyfish`
>    - {% icon version%} *"Confidence threshold"*: `0.5`
>    - {% icon version %} *"Video frame stride"*: `5` (default)
>    - {% icon param-toggle %} *"Show bounding boxes on annotated output"*: `Yes` (default)
>    - {% icon param-toggle %} *"Normalize outputs?"*: `No` (default)
>
>    > <tip-title>How to write a good prompt?</tip-title>
>    >
>    > The text prompt should describe the object to segment in **English**, using simple and precise terms.
>    > To detect multiple classes at once, separate them with commas:
>    > `jellyfish, shrimp, fish`
>    >
>    > Avoid overly vague descriptions like `animal` if you are specifically looking for a jellyfish.
>    > You can also use more descriptive prompts like `small blue fish`, but results may vary.
>    {: .tip}
>

> > 2. Click **Execute**

Reviewer 2: ```suggestion
> 2. Click **Run Tool**
```

>    - {% icon param-toggle %} *"Show bounding boxes on annotated output"*: `Yes` (default)
>    - {% icon param-toggle %} *"Normalize outputs?"*: `No` (default)
>
>    > <tip-title>Understanding the video parameters</tip-title>
>    >
>    > {% icon version %} *"Video frame stride"*: determines how often frames are analysed. A stride of `5` means one frame in every five is processed.
>    > - **Low stride (1–3)**: more precise analysis, but longer processing time
>    > - **High stride (10–30)**: faster processing, useful for long videos where objects move slowly
>    >
>    > {% icon param-select %} *"Video quality"*: controls the quality of the annotated output video, with no impact on processing speed or annotations.
>    >
>    > {% icon param-select %} *"COCO output mode"*: controls how COCO annotations are generated.
>    > - `Annotate the video`: one COCO entry per frame, referencing the video file (default)
>    > - `Annotate extracted frames`: saves frames as JPGs with one COCO entry per image — useful for pre-processing, for example with the {% tool [AnyLabeling Interactive](interactive_tool_anylabeling) %} tool as shown in the Moorev tutorial
>    >
>    {: .tip}
>

> > 2. Click **Execute**

Reviewer 2: ```suggestion
> 2. Click **Run Tool**
```
